### ISSUE TITLE
Assign Groups to custom roles is duplicating

### ISSUE THREAD

**Issue by mattmattox** [rancher/rancher#28647]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- Create a custom role in Rancher UI
- Use the Rancher API to assign that custom role to a group
- Run that Rancher API again

**Result:**
- Multiple role assignments for that group will be created. 

**Other details that may be helpful:**
- If you delete the one of the rolebinding the other one will be left behind and the group will still have the permissions

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.3.6
- Installation option (single install/HA): HA

**Comment by vincent99** [comment]
What are the actual API calls you're making?  How far apart?

There are no "unique indexes" in etcd other than the name, so it's not really possible to fully guarantee that you can't enter duplicates of any other data.

There's also nothing "wrong" with having two roles bindings for the same role and user (though our PRTB/CRTBs don't need to allow/create that situation in the actual RoleBindings they create).

Even if you prevent duplicates, a user might have a particular permission through multiple bindings to different groups and/or roles.  So a single delete on its own never guarantees you _actually_ took some permission away from somebody.

**Comment by mattmattox** [comment]
@vincent99 This API call is being made as part of a cluster automation job which is used to create and update all our clusters. Because of this, this call is being made multiple times per day as nodes are added or removed.

Would it be possible to look for an existing rolebinding that has the same role and user then reject the request?

**Comment by vincent99** [comment]
Not really; that turns it into a race where you get duplicates in hard-to-reproduce cases where they're posted at almost the same time. Both calls look, both see no entry exists and then both add one.

If that is "close enough" and ok for the user's use case, then great, their automation can perform that check itself today.  But if _we_ say uniqueness is enforced by our API, then it needs to _always_ be true.

The only thing that can be enforced at the k8s/database level to be unique is a resource's name.  And that doesn't really help here, because even if you somehow encoded it into the name, the bindings are also _editable_.  So the person/group/role involved can change (while the name of an existing resource can't).
